python/ray/serve/_private/http_state.py

```diff
@@ -404,3 +434,17 @@ async def ensure_http_route_exists(self, endpoint: EndpointTag, timeout_s: float
                 for proxy in self._proxy_states.values()
             ]
         )
+
+    def update_draining_flags(self, active_nodes: Set[str]):
+        """Update the draining states of all HTTP proxies.
+
+        Given a set of active nodes, set the draining flag of all HTTP proxies, except
+        for head node. Head node will never be draining.
+        """
+        for node_id, proxy_state in self._proxy_states.items():
+            # Head node will always be draining.
+            if node_id == self._head_node_id:
+                continue
+
+            draining = node_id not in active_nodes
+            proxy_state.set_draining_flag(node_id=node_id, draining=draining)
```

Comment:
This will cause us to constantly broadcast this flag to every proxy in the cluster on every iteration cycle. This will cause quite a bit of additional traffic in the cluster.

We have a long polling system set up that can be easily reused for this purpose, which I think would be preferable. Look for `LongPollHost` and `LongPollClient`. What you'll need to do is:

- Pass the long poll host into http_state and broadcast a dict of {node_ip: draining_flag}. You can dedup the updates by keeping a cached version of this dict in http_state (it's ok to always re-broadcast when the controller recovers from a crash).
- In the long poll client in http_proxy, add a new listener that listens for the draining map and updates its flag accordingly.

This should make it easy to avoid making unnecessary duplicate actor calls to the proxies while gracefully handling failure recovery (when a proxy restarts it will re-request the draining map).

Reply:
Thanks for the suggestion, will look into how this works!

Reply:
Figured out long poll. Implemented as the following:
- controller keeps a set of `active_nodes` and have an instance of `LongPollHost`
- during each `update()` call, controller will generate a set of `active_nodes`
- if the new `active_nodes` set changes from the current state, `notify_changed()` will be called
- both http proxy state and http proxy each has an instance of `LongPollClient` listening to the change on the `ACTIVE_NODES` namespace
- http proxy state has a callback to set the `self_draining` flag on itself. And the `update()` call will use this to determine the new state
- http proxy also has a callback to set the `self._draining` flag on itself. And there are logic to log flag changes and if we need to reset the dummy object reference (this is for minimizing `ray.put()` calls only when necessary). The flag will also continue to be used to determine the response of `/-/routes` and `/-/healthz`